### ISSUE TITLE
Support terminating `ThreadPool` without executing remaining jobs.

### ISSUE THREAD

**Issue by jonhoo** [rayon-rs/rayon#544]
There is currently only one way to terminate a `ThreadPool`; by using `drop`. According to the documentation:

> When the `ThreadPool` is dropped, that's a signal for the threads it manages to terminate, they will complete executing any remaining work that you have spawned, and automatically terminate.

While this is normally the desired behavior, there are cases where you might want to *not* execute remaining jobs, and instead just have the workers drop that work and return. An example of this would be if you write a open-loop benchmark that issues requests using a `ThreadPool`, and the user has specified some maximum time for the benchmark to run. If the threads aren't keeping up with the generated load, the queue will keep growing, and dropping the pool after the user-specified timeout may take a very long time.

It would be useful for situations like this to have a `cancel` method on `ThreadPool` that allows instructing the workers to finish just the tasks they are currently executing, and then exit, without working through the entire backlog. Ideally, `cancel` would also return how many jobs were left in the queue, but this is mostly useful for debugging purposes.

**Comment by jonhoo** [comment]
I guess one workaround for this would be to `mem::forget` the pool and have `main()` exit, but that seems pretty severe, and will mean that cleanup code in `drop`s and worker exit functions won't be run. It also gives no way of detecting how long the backlog is. I suppose the workers could maintain this count themselves (when queuing a job, atomically increment counter, and inside every job atomically decrement that same counter), but that would add some non-trivial complexity..

**Comment by cuviper** [comment]
I'm not sure how we can tell the difference between "backlog" and stuff that's currently executing.  For instance, a parallel iterator will recursively call `join`, queuing jobs, but those still need to execute before that calling frame can return.  I think the only things that could easily be discarded are from the static `spawn` and `ThreadPool::spawn`.

**Comment by jonhoo** [comment]
Yeah, I guess this only really makes sense for those who use `ThreadPool` exclusively as a `ThreadPool` for spawning jobs on (i.e., using [`ThreadPool::spawn`](https://docs.rs/rayon/1.0.0/rayon/struct.ThreadPool.html#method.spawn)). That seems like a decently common use-case though, and in those cases I think it makes a lot of sense to be able to do a clean early exit?

**Comment by jonhoo** [comment]
One thing that may make it easier to approximate this is to provide a `job_handler` for a `ThreadPool` which takes a closure that is executed just before running each job. This would be analogous to the existing `start_handler` and `exit_handler`, except for every job. How feasible is it to add something like this?
